EIPS/eip-5164.md

```diff
@@ -26,203 +26,286 @@ The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL
 
-`CrossChainExecutor`s MUST NOT execute a batch of calls more than once.
+`MessageExecutor`s MUST NOT successfully execute a message more than once.
+
+`MessageExecutor`s MUST revert the transaction when a message fails to be executed allowing the message to be retried at a later time.
 
 **Calldata**
 
-`CrossChainExecutor`s MUST append the ABI-packed (`nonce`, `sender`) to the calldata for each call being executed. It allows the receiver of the call to check the true `sender` of the transaction and use the `nonce` to apply any transaction ordering logic.
+`MessageExecutor`s MUST append the ABI-packed (`fromChainId`, `from`) to the calldata for each message being executed. This allows the receiver of the message to verify the cross-chain sender and the chain that the message is coming from.
 
 ```solidity
-interface CrossChainExecutor {
-  bytes calldata = abi.encode(Call.data, nonce, sender); // Can also use abi.encodePacked
+interface MessageExecutor {
+  bytes calldata = abi.encodePacked(fromChainId, from);
 }
 ```
 
 ```yaml
 - name: calldata
   type: bytes
   inputs:
-    - name: Call.data
-      type: bytes
-    - name: nonce
+    - name: fromChainId
       type: uint256
-    - name: sender
+    - name: from
       type: address
 ```
 
 #### Error handling
 
-**CallsAlreadyExecuted**
+**MessageAlreadyExecuted**
 
-`CrossChainExecutor`s MUST revert if a batch of calls has already been executed and SHOULD emit `CallsAlreadyExecuted` custom error.
+`MessageExecutor`s MUST revert if a messageId has already been executed and SHOULD emit a `MessageIdAlreadyExecuted` custom error.
 
 ```solidity
-interface CrossChainExecutor {
-  error CallsAlreadyExecuted(
-    uint256 nonce
+interface MessageExecutor {
+  error MessageIdAlreadyExecuted(
+    bytes32 messageId
   );
 }
 ```
 
-**CallFailure**
+**MessageFailure**
 
-`CrossChainExecutor`s SHOULD revert with a `CallFailure` error if a call fails.
+`MessageExecutor`s MUST revert if an individual message fails and SHOULD emit a `MessageFailure` custom error.
 
 ```solidity
-interface CrossChainExecutor {
-  error CallFailure(
-    uint256 callIndex,
+interface MessageExecutor {
+  error MessageFailure(
     bytes errorData
   );
 }
 ```
 
-#### Events
+**MessageBatchFailure**
 
-**ExecutedCalls**
+`MessageExecutor`s MUST revert the entire batch if any message in a batch fails and SHOULD emit a `MessageBatchFailure` custom error.
 
-`ExecutedCalls` MUST be emitted once calls have been executed.
+```solidity
+interface MessageExecutor {
+  error MessageBatchFailure(
+    uint256 messageIndex,
+    bytes errorData
+  );
+}
+```
+
+#### MessageExecutor Events
+
+**MessageIdExecuted**
+
+`MessageIdExecuted` MUST be emitted once a message or message batch has been executed.
 
 ```solidity
-interface CrossChainExecutor {
-  event ExecutedCalls(
-    CrossChainRelayer indexed relayer,
-    uint256 indexed nonce
+interface MessageExecutor {
+  event MessageIdExecuted(
```

Comment:
Suggest renaming to "MessageExecuted" and excluding the `fromChainId`

Indexers will see the "MessageDispatched" first on the sending chain, so they'll already have all the info

Reply:
The `fromChainId` is useful in frontend/sdk use cases where you have the `MessageExecuted` event and want to look up the `MessageDispatched` event back on the origin chain. That way you don't need to check against every network you support when things like subgraphs, RPCs, etc. are divided by network.